docs/design/coreclr/botr/type-loader.md

```diff
@@ -172,11 +172,31 @@ These are valid types and apparently `A` depends on `B` and `B` depends on `A`.
+While executing the code in `ClassLoader::CreateTypeHandleForTypeDefThrowing` before the call to `MethodTableBuilder::BuildMethodTableThrowing` no logic can rely on the `MethodTable` of the type that is being loaded.
+
+This has various implications, but the most obvious is that the base type of the type being loaded and any associated interfaces or field types cannot be loaded past `CLASS_LOAD_APPROXPARENTS` without creating a risk of triggering a `TypeLoadException`. Exceptions to this rule exist, and are necessary to actually implement the type loading process, but generally should be avoided, as they cause behavior which does not match the ECMA specification.
+
+#### 2.1.1.2 Code within `ClassLoader::DoIncrementalLoad`
+Code that runs during `DoIncrementalLoad` is generally allowed to require a class load to either the level that is being incrementally loaded to, OR the level at which the type being loaded is already at. The distinction here is that if the relationship between the types is circular, or non-circular. Circular relationships such as the relationship of a type to its type parameters can only be loaded to a level below the desired load level. Non-circular relationships can be required to be loaded to the load level that the incremental operation will eventually reach.
+
+For instance, the relationship of a type to its base type is non-circular, as a type cannot transitively be its own exact base type.
+However, the relationship of a type to the instantiation arguments of its base type can be circular, as a type `A` can be derived from a type `B<A>`.
+
+Code that runs in `ClassLoader::DoIncrementalLoad` follows a fairly straightforward pattern where code can depend on types being loaded a specific load level, and when the incremental load process completes at a given level, the type being loaded is incremented in load level.
+
+#### 2.1.1.3 Code within `PushFinalLevels`
+The final two levels of type loading are handled via `PushFinalLevels` which follows a different set of rules. `PushFinalLevels` runs code which in order to raise the level can only depend on other types being loaded to a level below the level that is desired. However, before marking the type as reaching a higher level, `PushFinalLevels` can require other types to also complete the `PushFinalLevels` algorithm to the new level. Only once all of the types are confirmed to have reached the new level can the entire set of types be marked as reaching the new level.
+
+### 2.1.2 Usage of load levels outside of the type loader.
+In the general case, its preferable to simply ignore load levels when not operating code that is part of the type loader, and to simply ask for fully loaded types. This should be the default, and always functionally correct choice. However, for performance reasons, it is possible to only require partially loaded types, which then requires the user of the type to ensure that their code does not have any dependencies on a fully loaded state.
```

Comment:
> their code does not have any dependencies on a fully loaded state.

Perhaps an example of information that requires a fully loaded state.

Reply:
This detail for that one gets to be a bit Meta... in that its typically not information on the type that one needs, but instead knowledge that a given type *CAN* be loaded to a fully loaded state. I don't have a good example, so I'll pass on this comment.

Reply:
That makes sense.